src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,138 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexToAdd(index);
+        Node<T> thatNode = new Node<>(value);
+        if (head == null) {
+            head = tail = thatNode;
+        } else if (index == 0) {
+            thatNode.next = head;
+            head = thatNode;
+        } else if (index == size) {
+            tail.next = thatNode;
+            tail = thatNode;
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            thatNode.next = prev.next;
+            prev.next = thatNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            add(listElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidator(index);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexValidator(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldNode = node.element;
+        node.element = value;
+        return oldNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidator(index);
+        T removedElement;
+        if (index == 0) {
+            removedElement = head.element;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> prev = getNodeByIndex(index - 1);
+            removedElement = prev.next.element;
+            prev.next = prev.next.next;
+            if (index == size - 1) {
+                tail = prev;
+            }
+        }
+        size--;
```

Comment:
move this code to method unlink() you should create this method (see common mistakes)